### ISSUE TITLE
Can we dim in a more intelligent way?

### ISSUE THREAD

**Issue by smichel17** [raatmarien/red-moon#122]
Okay, here's the thought: There's two ways to dim the screen: Turning down the backlight and using the filter. Given the choice, I think I would always rather start with the backlight, and only engage the filter when the backlight reaches 0. 

But, this is actually a really hard problem to solve.

- What kind of control over the backlight would we want to give? A single slider for dim that first controls the backlight and then controls the filter once the backlight reaches zero? Separate sliders? Some type of conversion where after the user changes the brightness on their device (can we detect that?), we'll lower the brightness a bit and turn down the filter's dim?
- Backlights of different devices may behave in different ways, so we might not be able to do a simple conversion from dim to backlight
- Users can adjust the backlight in real time, independently of the filter. I don't know how valuable others may find this.
- We need to ask for an additional permission before we can adjust the backlight, but we'd want the app to behave consistently. How would we handle each case?

I am not planning on working on this any time soon, but I am interested in anyone's feedback. What do you find useful or not about Red Moon's current brightness/dim controls? Is there anything you wish Red Moon could do, or that you never want it to do?

**Comment by mirh** [comment]
Also, why not hooking into KCAL, or using a driver like CF.lumen?

**Comment by smichel17** [comment]
#20

**Comment by smichel17** [comment]
Currently thinking it might make sense to replace `lower brightness` with settings to control the system brightness. A couple ways we could do that:

1. An app-wide toggle for whether Red Moon should control system brightness.
  - If it's enabled, add a checkbox setting and another slider to each filter, to control the brightness level and whether automatic brightness is enabled
2. Same as 1, but the toggle for controlling system brightness is per-filter.
3. Same as 2, but combine the two toggles into a list, ie `Control brightness: [Off / Fixed / Auto]`
  - If it's set to `Off`, the brightness slider disappears.

**Comment by smichel17** [comment]
Here's another one. It's more clever than the others. Maybe too clever.

- Add an app-wide toggle for whether to control the backlight
  - Allow toggling this from the notification
- Add a per-filter toggle to control whether automatic brightness is enabled
- Make the `dim` slider also control the system brightness.
  - If it's above halfway, it'll control the backlight brightness.
  - At halfway, when the backlight hits the minimum, it'll start to dim the screen using the filter.

**Comment by mirh** [comment]
> A toggle to control whether automatic brightness is enabled

If any, it should be natural to just stick to it if user had it enabled. 
~~I mean.. Provided ALS curves are exposed to programs I guess?~~ Ok [wrong](https://androidforums.com/threads/info-guide-alc-automatic-brightness-and-you-and-me-and-your-cat.841588/) assumption. And I guess adaptive brightness introduced in L further complicates the situation. 

...

In other news I found https://github.com/fython/Blackbulb

> If it's above halfway, it'll control the backlight brightness.

I'd say 1/4 imo. Or at least because on my phone minimum brightness is already pretty low.

**Comment by smichel17** [comment]
> If any, it should be natural to just stick to it if user had it enabled.

Right now, `Lower brightness`, if enabled, turns automatic\* brightness off and sets the brightness level to 0.

\*ie, adaptive brightness. I use the names I do because that's what they're called in the code.


> In other news I found https://github.com/fython/Blackbulb

Thank you for reminding me of this. I'd been meaning to look into how it's able to work without overlay permission (turns out, the filter is masquerading as a temporary notification).

> I'd say 1/4 imo. Or at least because on my phone minimum brightness is already pretty low.

1/2 was just an example; *If* I end up going this route, I'll play around with it to see what value feels good.

**Comment by smichel17** [comment]
Stray thought: Would it make sense to combine this with the button backlight setting? It would be nice if that setting had *some* effect on all devices, as I don't really like a useless setting sitting around.

**Comment by smichel17** [comment]
I updated the top comment with a proposal.

**Comment by smichel17** [comment]
Copying the reply from @wavexx in #191 so I don't forget it:

> I'm not sure all of this is needed. I agree that changing brightness
> should dim/[un]dim before changing the backlight when going under 0.
> 
> Ideally "dim" just doubles the dynamic range of the brightness, so that
> 0-50% is from maximum to zero dim (keeping backlight to 0), and 50-100%
> is zero backlight to 100% (keeping dim at 0).
> 
> If you could remap the system backlight control to this, then switching
> filters would be trivial (the current brightness is absolute). I have no
> experience with android development, so I cannot comment if that's
> actually doable.
> 
> However, when I switch off the filter, just restore whatever brightness
> was set before turning it on.
> 
> My specific use-case is super-simple: I just want darker, at night.
> I just use "dim", with the timer.

I do agree that this would be ideal. Unfortunately, it is impossible to change the behavior of the system brightness control. I also don't know of any way for an app to be notified when the user changes the brightness, although we can periodically check what the brightness is and react accordingly -- this might have an impact on battery usage, but we can minimize that if `Pause in excluded apps` is enabled.

When I came up with the proposal, I was thinking, "What are the options with the controls we have? Which of those is the best?" This comment got me thinking, "What's the ideal? How close can we get to that?" It was a very useful change of mindset and has given me some new ideas -- thanks :)

**Comment by wavexx** [comment]
On Thu, Sep 28 2017, smichel17 wrote:
> I do agree that this would be ideal. Unfortunately, it is impossible
> to change the behavior of the system brightness control. I also don't
> know of any way for an app to be notified when the user changes the
> brightness, although we can periodically check what the brightness is
> and react accordingly -- this might have an impact on battery usage,
> but we can minimize that if `Pause in excluded apps` is enabled.

Could you get at least increase/decrease brightness events? (as opposed
to sliding the brightness control, which I assume generates absolute
values).

> When I came up with the proposal, I was thinking, "What are the
> options with the controls we have? Which of those is the best?" This
> comment got me thinking, "What's the ideal? How close can we get to
> that?" It was a very useful change of mindset and has given me some
> new ideas -- thanks :)

But if you *did* include the backlight level in the filter settings, as
opposed to just "lower to 0", it would be just a setting like the other.
When changing the backlight, just update the preset value as well.

I never had more than one filter, but if I had more in addition, I
_wouldn't_ want the current brightness to influence my "dark reading"
preset where I want the backlight to be always zero.

**Comment by smichel17** [comment]
> Could you get at least increase/decrease brightness events? (as opposed to sliding the brightness control, which I assume generates absolute values).

Nope, there's no interface at all for apps to interact with the backlight brightness that I'm aware of, aside from getting and setting the backlight values (which is done the same way you'd change any other system setting, like toggling the wifi).

> But if you *did* include the backlight level in the filter settings, as opposed to just "lower to 0", it would be just a setting like the other. When changing the backlight, just update the preset value as well.

If possible I'd like to integrate backlight controls without increasing the number of settings. At its core, Red Moon has a pretty simple function and its design should reflect that. There are ways to do what you want without adding settings, though (for example, making the `Lower backlight` option multiple choice and/or combining it with `Hardware button backlight`).

> I never had more than one filter, but if I had more in addition, I _wouldn't_ want the current brightness to influence my "dark reading" preset where I want the backlight to be always zero.

Noted. I think the current proposal is better than any previous version, but not good enough to build, yet.

**Comment by mirh** [comment]
> Nope, there's no interface at all for apps to interact with the backlight brightness that I'm aware of

That sounds something the kernel would care/handle
https://github.com/torvalds/linux/blob/master/include/linux/backlight.h any idea?

**Comment by smichel17** [comment]
Maybe it's possible with root? Such things are typically undocumented. I also haven't read 100% of the android documentation so it's possible there's a callback I missed somewhere, but it seems unlikely

**Comment by wavexx** [comment]
On Fri, Sep 29 2017, mirh wrote:
> That sounds something the kernel would care/handle
> https://github.com/torvalds/linux/blob/master/include/linux/backlight.h any idea?

Under vanilla linux, yes. Unified leds and backlight control is
available on KMS drivers (and some others) via /proc/. Events (user and
system driven) are also emitted via ACPI.

However, this might tell very little on what android actually does/has
under the hood.

**Comment by mirh** [comment]
ARM certainly gives no fucks to ACPI...
Though _something else_ in the kernel might or might not be exposed

Again, I guess this is yet another time one of those "perfect with root - horrible pretension to be even working without" situations

**Comment by AdamNiederer** [comment]
> Nope, there's no interface at all for apps to interact with the backlight brightness that I'm aware of

I know [Lux](https://play.google.com/store/apps/details?id=com.vito.lux) manages to do it; the brightness slider is entirely disabled when Lux is enabled, and one can control the system brightness and overlay degree from the notification.

I'm pretty sure the core mechanism of action doesn't require any additional permissions, either. It asks for usage access for some unrelated features.

**Comment by mirh** [comment]
I'm not sure what you actually meant with "entirely disabled" (always stuck on the same setting, is not technically "disabled"), but yest there is [this stupid](https://developer.android.com/reference/android/provider/Settings.System#screen_brightness) thing since the first version of android. 

It's not what dev was talking about, if you just proceeded reading the reminder of the following sentence. 
Could instead [this](https://stackoverflow.com/questions/46678412/on-android-how-to-read-screen-brightness-or-auto-adjustment-factor-when-system-a#comment80316011_46680365) or [this](https://www.androidheadlines.com/2018/03/android-p-illustrates-real-time-adaptive-brightness-changes.html) be of any help?
